### PR TITLE
feat: stack consumer for use with LSP visitor

### PR DESCRIPTION
**Pull request by LoZander**
This PR adds a new consumer for the LSP visitor. This stack consumer adds everything visited to a stack where the head is the last element visited.

Related to #8639

### CODE REVIEW

File: main/src/ca/uwaterloo/flix/api/lsp/StackConsumer.scala [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+case class StackConsumer() extends Consumer {
+  private var stack: List[AnyRef] = Nil
+
+  private def push(x: AnyRef): Unit = {
+    stack = x :: stack
+  }
+
+  /**
+    * @return [[Option]] containing the top element from the stack of AST elements that the consumer has collected.
+    *         Returns [[Null]] if the stack is empty.
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/api/lsp/StackConsumer.scala:38]: Null -> none?

File: main/src/ca/uwaterloo/flix/api/lsp/StackConsumer.scala [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+    */
+  def getHead: Option[AnyRef] = stack.headOption
+
+  /**
+    * @return The stack of AST elements that the consumer has collected.
+    */
+  def getStack: List[AnyRef] = stack
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/api/lsp/StackConsumer.scala:45]: I suggest to keep getStack, but to drop getHead and getTail. You have direct access to those operations on the list anyway.

File: main/src/ca/uwaterloo/flix/api/lsp/StackConsumer.scala (hunk 1 above)
Comment (LoZander) [reply main/src/ca/uwaterloo/flix/api/lsp/StackConsumer.scala:38]: oh yeah, ofc. Brain fart on my end ^^'

File: main/src/ca/uwaterloo/flix/api/lsp/StackConsumer.scala [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+import ca.uwaterloo.flix.api.lsp.Visitor.Consumer
+import ca.uwaterloo.flix.language.ast.Ast.{AssocTypeConstructor, Derivation, Derivations, EqualityConstraint, TraitConstraint}
+import ca.uwaterloo.flix.language.ast.Type
+import ca.uwaterloo.flix.language.ast.TypedAst.Pattern.Record.RecordLabelPattern
+import ca.uwaterloo.flix.language.ast.TypedAst.*
+import ca.uwaterloo.flix.language.ast.shared.Annotation
+import ca.uwaterloo.flix.language.ast.shared.SymUse.*
+
+/**
+  * Consumer that collects every visited AST nodes on a stack where the head is the last element visited.
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/api/lsp/StackConsumer.scala:27]: A bit more information about what is actually computed.

File: main/src/ca/uwaterloo/flix/api/lsp/StackConsumer.scala (hunk 2 above)
Comment (LoZander) [reply main/src/ca/uwaterloo/flix/api/lsp/StackConsumer.scala:45]: true

Comment (LoZander) [reply main/src/ca/uwaterloo/flix/api/lsp/StackConsumer.scala:45]: i suppose those only really make sense if you specifically didn't want to expose the entire list (for some reason)

Comment (magnus-madsen) [reply main/src/ca/uwaterloo/flix/api/lsp/StackConsumer.scala:45]: Yes, so kill them and then I can merge.

Comment (LoZander) [reply main/src/ca/uwaterloo/flix/api/lsp/StackConsumer.scala:45]: yep

File: main/src/ca/uwaterloo/flix/api/lsp/StackConsumer.scala (hunk 3 above)
Comment (LoZander) [reply main/src/ca/uwaterloo/flix/api/lsp/StackConsumer.scala:27]: yep

**Comment by LoZander** [comment]
Oh, bad commit names, my bad. I also removed the redundant functions as discussed

**Review by magnus-madsen — APPROVED** [review]

**Pull request merged**